src/main/java/net/sf/rails/ui/swing/GameUIManager.java

```diff
@@ -1267,7 +1275,37 @@ public void run() {
         return currentGuiPlayerNames;
     }
 
-   public class PlayerOrderView implements Observer {
+    public void saveLogs() {
+        LoggerContext lc = (LoggerContext) LoggerFactory.getILoggerFactory();
+
+        CyclicBufferAppender<?> buffer = (CyclicBufferAppender<?>) lc.getLogger(Logger.ROOT_LOGGER_NAME).getAppender("buffer");
+        if ( buffer == null ) {
+            log.warn("Unable to find cyclic buffer appender");
+            return;
+        }
+
+        PatternLayout layout = new PatternLayout();
+        layout.setContext(lc);
+        layout.setPattern("%d{MM-dd-yyyy:HH:mm:ss.SSS} [%thread] %-5level %logger{10}->%method\\(\\):%line - %msg%n");
+        layout.start();
```

Comment:
What is the reason that we can't use our "default" approach `private static final Logger log = LoggerFactory.getLogger(GameUIManager.class);` here?

Reply:
Because we're not actually creating a logger-- we are fetching a reference to the logger factory, getting a specific appender configured in logback.xml named "buffer" (which is a circular buffer), and then creating a pattern layout to format the log data (in theory you could configure the pattern in a dummy appender in logback.xml and use that but not sure of the value in that).

Finally we iterator over each of the log events currently contained in the buffer appender and write them out to a file.

`Logger log = LoggerFactory...` creates a new log appender for use in generating log events. Our code here is for reading previously created log events and flushing them to a new file on the fly.